polaris-core/src/main/java/org/apache/polaris/core/persistence/pagination/Page.java

```diff
@@ -18,25 +18,99 @@
 
-  public Page(PageToken pageToken, List<T> items) {
-    this.pageToken = pageToken;
+  private Page(PageToken request, @Nullable Token nextToken, List<T> items) {
+    this.request = request;
+    this.nextToken = nextToken;
     this.items = items;
   }
 
   /**
-   * Used to wrap a {@link List<T>} of items into a {@link Page <T>} when there are no more pages
+   * Builds a complete response page for the full list of relevant items. No subsequence pages of
+   * related data exist.
    */
   public static <T> Page<T> fromItems(List<T> items) {
-    return new Page<>(new DonePageToken(), items);
+    return new Page<>(PageToken.readEverything(), null, items);
+  }
+
+  /**
+   * Produces a response page by consuming the number of items from the provided stream according to
+   * the {@code request} parameter. Source items can be converted to a different type by providing a
+   * {@code mapper} function. The page token for the response will be produced from the request data
+   * combined with the pointer to the next page of data provided by the {@code dataPointer}
+   * function.
+   *
+   * @param request defines pagination parameters that were uses to produce this page of data.
+   * @param items stream of source data
+   * @param mapper converter from source data types to response data types.
+   * @param tokenBuilder determines the {@link Token} used to start the next page of data given the
+   *     last item from the previous page. The output of this function will be available from {@link
+   *     PageToken#value()} associated with the request for the next page.
+   */
+  public static <R, T> Page<R> mapped(
+      PageToken request, Stream<T> items, Function<T, R> mapper, Function<T, Token> tokenBuilder) {
+    List<R> data;
+    T last = null;
+    if (!request.paginationRequested()) {
+      // short-cut for "no pagination"
+      data = items.map(mapper).collect(Collectors.toList());
+    } else {
+      data = new ArrayList<>(request.pageSize().orElse(10));
+
+      Iterator<T> it = iterator(items.spliterator());
+      int limit = request.pageSize().orElse(Integer.MAX_VALUE);
+      while (it.hasNext() && data.size() < limit) {
+        last = it.next();
+        data.add(mapper.apply(last));
+      }
+
+      // Signal "no more data" if the number of items is less than the requested page size or if
+      // there is no more data.
+      if (data.size() < limit || !it.hasNext()) {
+        last = null;
+      }
+    }
+
+    return new Page<>(request, tokenBuilder.apply(last), data);
+  }
+
+  public List<T> items() {
+    return items;
+  }
+
+  /**
+   * Returns a page token in encoded form suitable for returning to API clients. The string returned
+   * from this method is expected to be parsed by {@link PageToken#build(String, Integer)} when
+   * servicing the request for the next page of related data.
+   */
+  public @Nullable String encodedResponseToken() {
+    return PageTokenUtil.encodePageToken(request, nextToken);
```

Comment:
Same comments from the previous PR -- encoding/decoding a given PageToken clearly seems to be within the purview of a given PageToken implementation